aquadoggo/src/service_manager.rs

```diff
@@ -62,29 +51,92 @@ where
     ///
     /// This gets automatically wrapped in a static, boxed and pinned function signature by the
     /// `async_trait` macro so we don't need to do it ourselves.
-    async fn call(&self, context: Context<D>, shutdown: Shutdown, tx: Sender<M>) {
+    async fn call(&self, context: D, shutdown: Shutdown, tx: Sender<M>) -> Result<()> {
         (self)(context, shutdown, tx).await
     }
 }
 
+/// Future which resolves as soon as the internal boolean flag gets flipped to "true".
+///
+/// The flag can be manually set with the `fire` method or automatically whenever the signal
+/// instance gets dropped. The latter is a trick to fire a signal when something panics.
+struct Signal(Arc<AtomicBool>);
+
+impl Signal {
+    /// Returns a new signal instance.
+    pub fn new() -> Self {
+        Self(Arc::new(AtomicBool::new(false)))
+    }
+
+    /// Sets the boolean flag to "true" which will resolve the future.
+    pub fn fire(&mut self) {
+        self.0.swap(true, Ordering::Relaxed);
+    }
+}
+
+impl Future for Signal {
+    type Output = ();
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        match self.0.load(Ordering::Relaxed) {
+            true => Poll::Ready(()),
+            false => {
+                cx.waker().clone().wake();
+                Poll::Pending
+            }
+        }
+    }
+}
+
+impl Clone for Signal {
+    fn clone(&self) -> Self {
+        Self(self.0.clone())
+    }
+}
+
+impl Drop for Signal {
+    fn drop(&mut self) {
+        // Fire signal whenever signal handler goes out of scope
+        self.fire();
+
+        // Drop object
+        drop(self)
+    }
+}
+
+// Service manager for orchestration of long-running concurrent processes.
+//
+// This manager offers a message bus between services for cross-service communication. It also
+// sends a shutdown signal to allow services to react to it gracefully.
+//
+// Stopped services (because of a panic, error or successful return) will send an exit signal which
+// can be subscribed to via the `on_exit` method. Usually stopped services indicate system failure
+// and it is recommended to stop the application when this events occurs.
```

Comment:
Cool, so this is the "stop everything if one service fails" pattern we just discussed in the previous PR?

Reply:
Correct, the service manager in itself does not enforce this, but it gives you the option.

The aquadoggo cli `main.rs` uses that pattern and moves over to shutdown as soon as CTRL+C has been pressed or a service stopped aka the exit signal fired.